compiler/src/main/java/com/squareup/anvil/compiler/ModuleMerger.kt

```diff
@@ -92,112 +84,107 @@ internal class ModuleMerger(
       // for replaced classes and the final result.
       .toList()
 
-    val excludedModules = (mergeAnnotation.getAnnotationValue("exclude") as? ArrayValue)
+    val excludedModules = (mergeAnnotation.annotation.getAnnotationValue("exclude") as? ArrayValue)
       ?.value
       ?.map {
         val argumentType = it.argumentType(module)
-        val classDescriptorForExclusion = argumentType.classDescriptor()
+        val excludedClass = argumentType.classDescriptor().toClassReference(module)
 
-        val contributesToAnnotation = classDescriptorForExclusion
-          .annotationOrNull(contributesToFqName)
-        val contributesBindingAnnotation = classDescriptorForExclusion
-          .annotationOrNull(contributesBindingFqName)
-        val contributesMultibindingAnnotation = classDescriptorForExclusion
-          .annotationOrNull(contributesMultibindingFqName)
-        val contributesSubcomponentAnnotation = classDescriptorForExclusion
-          .annotationOrNull(contributesSubcomponentFqName)
+        val contributesToAnnotation = excludedClass.annotationOrNull(contributesToFqName)
+        val contributesBindingAnnotation = excludedClass.annotationOrNull(contributesBindingFqName)
+        val contributesMultibindingAnnotation =
+          excludedClass.annotationOrNull(contributesMultibindingFqName)
+        val contributesSubcomponentAnnotation =
+          excludedClass.annotationOrNull(contributesSubcomponentFqName)
 
         // Verify that the replaced classes use the same scope.
-        val scopeOfExclusion = contributesToAnnotation?.scope(module)
-          ?: contributesBindingAnnotation?.scope(module)
-          ?: contributesMultibindingAnnotation?.scope(module)
-          ?: contributesSubcomponentAnnotation?.parentScope(module)
-          ?: throw AnvilCompilationException(
-            thisDescriptor,
-            "Could not determine the scope of the excluded class " +
-              "${classDescriptorForExclusion.fqNameSafe}."
+        val scopeOfExclusion = contributesToAnnotation?.scopeOrNull()
+          ?: contributesBindingAnnotation?.scopeOrNull()
+          ?: contributesMultibindingAnnotation?.scopeOrNull()
+          ?: contributesSubcomponentAnnotation?.parentScope()
+          ?: throw AnvilCompilationExceptionClassReference(
+            message = "Could not determine the scope of the excluded class " +
+              "${excludedClass.fqName}.",
+            classReference = excludedClass
           )
 
-        if (scopeOfExclusion.fqNameSafe != scopeFqName) {
-          throw AnvilCompilationException(
-            thisDescriptor,
-            "${thisDescriptor.fqNameSafe} with scope $scopeFqName wants to exclude " +
-              "${classDescriptorForExclusion.fqNameSafe} with scope " +
-              "${scopeOfExclusion.fqNameSafe}. The exclusion must use the same scope."
+        if (scopeOfExclusion.fqName != scopeFqName) {
```

Comment:
Prefer comparing `scopeOfExclusion` to `scope` instead of using the FqNames. We implement equals() and get more control this way.